mmpy_bot/plugins/base.py

```diff
@@ -88,21 +70,213 @@ async def call_function(
+    be displayed in the channel where it is requested.
+    """
+
+    def __init__(
+        self,
+        direct_help: bool = True,
+    ):
+        super().__init__()
+        self.direct_help: bool = direct_help
+
+    def initialize(
+        self,
+        driver: Driver,
+        plugin_manager: PluginManager,
+        settings: Settings,
+    ):
+        super().initialize(driver, plugin_manager, settings)
+
+        if self.settings.RESPOND_CHANNEL_HELP:
+            self.help = listen_to("^!help$")(self.help)
+
+    def get_help_string(self, message: Message) -> str:
+        def custom_sort(rec):
+            return (
+                rec.metadata.get("category", ""),  # No categories first
+                rec.help_type,
+                rec.pattern.lstrip("^[(-"),
+            )
+
+        help_function_info = sorted(self.get_help(message), key=custom_sort)
+
+        string = "### The following functions have been registered:\n\n"
+        string += "###### `(*)` require the use of `@botname`, "
+        string += "`(+)` can only be used in direct message\n"
+        old_category = None
+
+        for h in help_function_info:
+            # If categories are defined, group functions accordingly
+            category = h.metadata.get("category")
+            if category != old_category:
+                old_category = category
+                category = "uncategorized" if category is None else category
+                string += f"Category `{category}`:\n"
+
+            cmd = h.metadata.get("human_description", h.pattern)
+            direct = "`(*)`" if h.direct else ""
+            mention = "`(+)`" if h.mention else ""
+
+            if h.help_type == "webhook":
+                string += (
+                    f"- `{cmd}` {direct} {mention} - (webhook) {h.function_docheader}\n"
+                )
+            else:
+                if not h.function_docheader:
+                    string += f"- `{cmd}` {direct} {mention}\n"
+                else:
+                    string += f"- `{cmd}` {direct} {mention} - {h.function_docheader}\n"
 
         return string
 
+    def get_help(self, message: Message):
+        """Obtain Help info from PluginManager.
+
+        Override this method if you need to customize which listeners will be included
+        in the help.
+        """
+        return self.plugin_manager.get_help()
+
+    @listen_to("^help$", needs_mention=True)
     async def help(self, message: Message):
-        """Prints the list of functions registered on every active plugin."""
-        self.driver.reply_to(message, self.get_help_string())
+        """Shows this help information."""
+        self.driver.reply_to(
+            message, self.get_help_string(message), direct=self.direct_help
+        )
+
+
+@dataclass
+class FunctionInfo:
+    help_type: str
+    location: str
+    function: Function
+    pattern: str
+    plugin_docheader: str
+    plugin_docfull: str
```

Comment:
Storing this information for every function in the plugin is redundant, and I believe these values are actually not used anywhere? Then I'd prefer removing them.